File: frontend/src/utils/money.ts
/**
 * Cents-safe money formatting. Money travels as integer cents everywhere —
 * never floats — so formatting just divides by 100 at the very edge.
 */

/** `$3.00` for 300 cents; "Free" when the price is null (a free share). */
export function formatCents(cents: number | null | undefined): string {
  if (cents == null) return 'Free';
  return `$${(cents / 100).toFixed(2)}`;
}

/** `$3.00/hr` or "Free" — for share prices and price breakdowns. */
export function formatRate(cents: number | null | undefined): string {
  if (cents == null) return 'Free';
  return `${formatCents(cents)}/hr`;
}

File: backend/src/main/java/com/spotshare/common/Money.java
package com.spotshare.common;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Instant;

/**
 * Cents-safe money math. Money travels as integer cents everywhere and is
 * only divided into dollars at the display edge — never floating point.
 */
public final class Money {

    private Money() {
    }

    /**
     * Prorated total for a parking period (spec §7): billable minutes ×
     * (hourly rate / 60), rounded HALF_UP to cents. A free share
     * ({@code hourlyRateCents == null}) costs nothing.
     */
    public static int proratedTotalCents(Integer hourlyRateCents, Instant arrival, Instant departure) {
        if (hourlyRateCents == null) {
            return 0;
        }
        long minutes = java.time.Duration.between(arrival, departure).toMinutes();
        return BigDecimal.valueOf(hourlyRateCents)
                .multiply(BigDecimal.valueOf(minutes))
                .divide(BigDecimal.valueOf(60), 0, RoundingMode.HALF_UP)
                .intValueExact();
    }
}
